src/components/ConditionsLayer/Flow/Flow.tsx

```diff
@@ -0,0 +1,235 @@
+import { MouseEvent, useCallback, useEffect, useState } from "react";
+import ReactFlow, {
+  Background,
+  Controls,
+  Edge,
+  Node,
+} from "react-flow-renderer";
+import ConditionNode, {
+  ConditionNodeTypes,
+} from "components/ConditionsLayer/Nodes/ConditionNode";
+import DashedEdge from "components/Edges/DashedEdge";
+import {
+  EdgeTypes,
+  initialEdges,
+  initialNodes,
+  NodeTypes,
+} from "./initialElements";
+import StartEndNode from "../Nodes/StartEndNode";
+import { StartEndNodeTypes } from "../Nodes/StartEndNode/StartEndNode";
+
+const nodeTypes = { condition: ConditionNode, startEnd: StartEndNode };
+const edgeTypes = {
+  dashed: DashedEdge,
+};
+
+interface FlowProps {
+  arePlaceholdersVisible: boolean;
+}
+
+const findLastColumn = (nodes: Node[]) => {
+  let lastCol = 0;
+  nodes.map(({ data }) => (data.col > lastCol ? (lastCol = data.col) : null));
+  return lastCol;
+};
+
+interface ColumnData {
+  col: number;
+  row: number;
+  id: string | null;
+}
+
+const findLastRow = (nodes: Node[]) => {
+  let maxRow = 0;
+  nodes.map(({ data }) => (data.row > maxRow ? (maxRow = data.row) : null));
+  return maxRow;
+};
+
+const lastElementInEachRow = (nodes: Node[]) => {
+  let colArr: ColumnData[] = [];
+  const preparedNodes = nodes.filter(
+    (item) => item.data.type !== ConditionNodeTypes.PLACEHOLDER
+  );
+
+  preparedNodes.map(({ data }) => {
```

Comment:
suggestion: you don't need to push to an array defined earlier, since the `.map` function returns a new array.

```
const colArr: ColumnData[] = preparedNodes.map(({data}) => {
    // some stuff here
    return {} // the return value is the stuff that will be "pushed" to the new array
});`